File: rmats-viz/backend/app/schemas/gene.py
"""
Gene schemas
============
Pydantic models for gene search results and stored gene entries.

``GeneEntry`` is the canonical representation used both in API responses
and in the ``mutated_genes`` JSONB column of the ``Analysis`` model.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class GeneEntry(BaseModel):
    """A resolved gene with its Ensembl stable identifier."""

    symbol: str = Field(..., description="HUGO gene symbol, e.g. 'BRCA1'")
    ensembl_id: str = Field(..., description="Ensembl stable gene ID, e.g. 'ENSG00000012048'")
    display: str = Field(..., description="Human-readable label: 'BRCA1 (ENSG00000012048)'")

    model_config = {"from_attributes": True}


class GeneSearchResponse(BaseModel):
    """Wrapper returned by the gene search endpoint."""

    query: str
    results: list[GeneEntry]
    total: int

File: rmats-viz/frontend/src/types/gene.ts
/**
 * Gene types
 * ===========
 * TypeScript interfaces for Ensembl gene entries used in autocomplete
 * and stored in analyses.
 */

/** A resolved gene entry with its Ensembl stable identifier. */
export interface GeneEntry {
  /** HUGO gene symbol, e.g. "BRCA1" */
  symbol: string;
  /** Ensembl stable gene ID, e.g. "ENSG00000012048" */
  ensembl_id: string;
  /** Human-readable label: "BRCA1 (ENSG00000012048)" */
  display: string;
}

/** Response from GET /api/v1/genes/search */
export interface GeneSearchResponse {
  query: string;
  results: GeneEntry[];
  total: number;
}
